File: tp-5/Set.hs
module Set 

    (Set, emptyS, addS, belongs, sizeS, removeS, unionS, setToList)
    
    where

    data Set a = Set [a] Int
    -- Inv-rep: * Si la lista del set se encuentra vacia entonces el int es 0.
    --          * El Int representa la cantidad de elementos de la lista.
    --          * El Int no puede ser < 0.


    emptyS :: Set a 
    emptyS = (Set [] 0)                                                         -- O(1)

    addS :: Eq a => a -> Set a -> Set a
    addS            a    (Set xs int) = if elem a xs
                                            then (Set xs int)
                                            else (Set (a : xs) (int + 1))       -- O(n) donde n son la cantidad de elementos del set.

    belongs :: Eq a => a -> Set a -> Bool   
    belongs            a    (Set xs int) = elem a xs                            -- O(n) donde n son la cantidad de elementos del set.

    sizeS :: Eq a => Set a -> Int
    sizeS            (Set xs int) = int                                         -- O(1)

    removeS :: Eq a => a -> Set a -> Set a
    removeS            a    (Set xs int) = if elem a xs
                                            then (Set (removeL a xs) (int - 1))
                                            else (Set xs int)                   -- O(n) donde n son la cantidad de elementos del set.

    removeL :: Eq a => a -> [a] -> [a]
    removeL            _    []  = []
    removeL            x    (y:ys) = if x == y 
                                        then ys
                                        else x : (removeL x ys)                 -- O(n) donde n son la cantidad de elementos de la lista.

    unionS :: Eq a => Set a -> Set a -> Set a
    unionS            (Set xs int) set = unionSL xs set                         -- O(n) por el llamado a unionSL de costo n.

    unionSL :: Eq a => [a] -> Set a -> Set a 
    unionSL    []     set   = set
    unionSL    (x:xs) set   = addS x (unionSL xs set)                           -- O(n) donde n son la cantidad de elementos de la lista.

    setToList :: Eq a => Set a -> [a]
    setToList            (Set xs int) = xs                                      -- O(1) 

    

File: tp-11/Set.cpp
#include <iostream>
using namespace std;
#include "Set.h"

struct SetSt
//Inv-rep: * cantidad son la cantidad de nodos desde el primero que se pueden recorrest hasta llegar a NULL.
//         * para cada NodoS dentro del SetSt su valor del campo "elem" no se repite en otro NodoS.
{
    int cantidad;
    NodoS* primero;
};


Set emptyS(){
    Set s = new SetSt;
    s->cantidad = 0;
    s->primero = NULL;
}

bool isEmptyS(Set s){
    return s->cantidad == 0;
}

bool belongsS(int x, Set s){        // Costo: O(n) donde n son la cantidad de elementos del set.
    NodoS* p = s->primero;
    for (int i = 0; i < s->cantidad && p->elem != x; i++)
    {
        p = p->siguiente;
    }
    return p != nullptr;
}

void AddS(int x, Set s){
    if (!belongsS(x, s))
    {
        NodoS* n = new NodoS;
        n->elem = x;
        n->siguiente = s->primero;
        s->primero = n;
        s->cantidad++;
    }
}

void RemoveS(int x, Set s){
    NodoS* p = s->primero;
    NodoS* a;
    for (int i = 0; i < s->cantidad && p->elem != x; i++)
    {
        a = p;
        p = p->siguiente;
    }
    if (p != nullptr)
    {
       a->siguiente = p->siguiente;
       s->cantidad--;
       delete p;
    }
}

int sizeS(Set s){
    return s->cantidad;
}

LinkedList setToList(Set s){
    NodoS* p = s->primero;
    LinkedList l = nil();
    for (int i = 0; i < s->cantidad; i++)
    {
        Cons(p->elem, l);
    }
    return l;
}

void DestroyS(Set s);